controllers/release/release_adapter_test.go

```diff
@@ -0,0 +1,421 @@
+			}, release)
+		}, time.Second*10).Should(HaveOccurred())
+
+		Eventually(func() bool {
+			pipelineRun, _ := adapter.getReleasePipelineRun()
+			return (pipelineRun == nil)
+		}, time.Second*10).Should(BeTrue())
+	})
+
+	It("ensures Finalizers are called and fails for a Release object when the PipelineRun was already finalized", func() {
+		// Call to EnsureFinalizersAreCalled should succeed if the Release was not set to be deleted
+		result, err := adapter.EnsureFinalizersAreCalled()
+		Expect(result.CancelRequest).To(BeFalse())
+		Expect(err).NotTo(HaveOccurred())
+
+		// Create a ReleasePipelineRun to ensure it gets finalized
+		result, err = adapter.EnsureReleasePipelineRunExists()
+		Expect(result.CancelRequest).To(BeFalse())
+		Expect(err).NotTo(HaveOccurred())
+
+		// Add finalizer
+		result, err = adapter.EnsureFinalizerIsAdded()
+		Expect(result.CancelRequest).To(BeFalse())
+		Expect(err).NotTo(HaveOccurred())
+
+		result, err = adapter.EnsureReleasePipelineStatusIsTracked()
+		Expect(result.CancelRequest).To(BeFalse())
+		Expect(err).NotTo(HaveOccurred())
+
+		release.SetDeletionTimestamp(&metav1.Time{Time: time.Now()})
+
+		// delete the pipelineRun before the release so it should throw an
+		// error on finalizing the release
+		pipelineRun, _ := adapter.getReleasePipelineRun()
+		Expect(k8sClient.Delete(ctx, pipelineRun)).Should(Succeed())
+		Expect(k8sClient.Delete(ctx, release)).Should(Succeed())
+
+		result, err = adapter.EnsureFinalizersAreCalled()
+		Expect(!result.CancelRequest && err != nil).To(BeTrue())
+	})
+
+	It("can create a ReleasePipelineRun", func() {
+		pipelineRun, err := adapter.createReleasePipelineRun(
+			releaseStrategy,
+			applicationSnapshot)
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(pipelineRun.Spec.ServiceAccountName).Should(ContainSubstring("test-account"))
+		Expect(pipelineRun.Spec.Params[0].Value.StringVal).Should(ContainSubstring("test-policy"))
+		Expect(k8sClient.Delete(ctx, pipelineRun)).Should(Succeed())
+	})
+
+	It("can finalize (delete) the releasePipelineRun just processed", func() {
+		result, err := adapter.EnsureFinalizersAreCalled()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		Eventually(func() bool {
+			result, err := adapter.EnsureFinalizerIsAdded()
+			return (!result.CancelRequest && err == nil)
+		}, time.Second*10).Should(BeTrue())
+
+		Eventually(func() bool {
+			result, err := adapter.EnsureReleasePipelineRunExists()
+			return (!result.CancelRequest && err == nil)
+		}, time.Second*10).Should(BeTrue())
+		result, err = adapter.EnsureReleasePipelineStatusIsTracked()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		err = adapter.finalizeRelease()
+		Expect(err).Should(Succeed())
+
+		Eventually(func() bool {
+			pipelineRun, _ := adapter.getReleasePipelineRun()
+			return (pipelineRun == nil)
+		}, time.Second*10).Should(BeTrue())
+	})
+
+	It("can return the pipelineRun from the release", func() {
+		Eventually(func() bool {
```

Comment:
No race condition AFAIK.

Reply:
I don't like having to use Eventually blocks. If you see them is because it was required by the test. But I am happy to crosscheck.

Reply:
Oudated. Not needed.